File: stacksearch-webapp/src/app/search-results/search.service.ts
import {Injectable} from "@angular/core";
import {Observable} from "rxjs";
import {Page} from "../model/page.model";
import {Question} from "../model/question.model";
import {HttpClient, HttpParams} from "@angular/common/http";

@Injectable()
export class SearchService {

  constructor(private httpClient: HttpClient) {}

  search(query: string, page: number, size: number): Observable<Page<Question>> {
    let params = new HttpParams().set('query', query).set('page', '' + page).set('size', '' + size);
    return this.httpClient.get<Page<Question>>('http://localhost:8080/search', {params});
  }
}

File: stacksearch-api/src/main/kotlin/absaliks/stacksearch/SearchService.kt
package absaliks.stacksearch

import absaliks.stacksearch.model.Page
import absaliks.stacksearch.model.Question
import absaliks.stacksearch.model.emptyPage
import org.springframework.core.ParameterizedTypeReference
import org.springframework.http.HttpMethod
import org.springframework.http.ResponseEntity
import org.springframework.stereotype.Service
import org.springframework.web.client.RestOperations

const val REMOTE_URL =
    "http://api.stackexchange.com/2.2/search?intitle={query}&page={page}&pagesize={size}&site=stackoverflow"

inline fun <reified T> typeRef() = object : ParameterizedTypeReference<T>() {}

@Service
class SearchService(private val restOperations: RestOperations) {

    fun searchQuestions(criteria: SearchCriteria): Page<Question> {
        val params = toStackExchangeParamMap(criteria)
        val responseEntity: ResponseEntity<Page<Question>> =
                restOperations.exchange(REMOTE_URL, HttpMethod.GET, null, typeRef<Page<Question>>(), params)

        return responseEntity.body ?: emptyPage()
    }

    fun toStackExchangeParamMap(criteria: SearchCriteria): HashMap<String, Any?> {
        return hashMapOf(
            "query" to criteria.query,
            "page" to criteria.page + 1,
            "size" to criteria.size
        )
    }
}
